v2/cmd/controller/app/setup.go

```diff
@@ -0,0 +1,352 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package app
+
+import (
+	"context"
+	"flag"
+	"math/rand"
+	"os"
+	"time"
+
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/benbjohnson/clock"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/rest"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/healthz"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/Azure/azure-service-operator/v2/identity"
+	"github.com/Azure/azure-service-operator/v2/internal/config"
+	"github.com/Azure/azure-service-operator/v2/internal/controllers"
+	"github.com/Azure/azure-service-operator/v2/internal/crdmanagement"
+	"github.com/Azure/azure-service-operator/v2/internal/genericarmclient"
+	. "github.com/Azure/azure-service-operator/v2/internal/logging"
+	asometrics "github.com/Azure/azure-service-operator/v2/internal/metrics"
+	armreconciler "github.com/Azure/azure-service-operator/v2/internal/reconcilers/arm"
+	"github.com/Azure/azure-service-operator/v2/internal/reconcilers/generic"
+	internalflags "github.com/Azure/azure-service-operator/v2/internal/util/flags"
+	"github.com/Azure/azure-service-operator/v2/internal/util/interval"
+	"github.com/Azure/azure-service-operator/v2/internal/util/kubeclient"
+	"github.com/Azure/azure-service-operator/v2/internal/util/lockedrand"
+	"github.com/Azure/azure-service-operator/v2/internal/version"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
```

Comment:
I'm not totally sold on the `app/setup.go` structure here. Did you take inspiration from someplace on this `app/...` structure?

Given this stuff is all very specific to `cmd/controller` I agree it makes sense to keep here and not put into something like `pkg/setup` or `internal/setup`... do we prefer `app/setup.go` or just having these functions defined in `main.go` (or `setup.go` living alongside `main.go`)?

Basically 100% in support of new `SetupControllerManager` function, but wondering what file it makes sense to go into.

Reply:
I took the inspiration from how [cert-manager](https://github.com/cert-manager/cert-manager/tree/master/cmd/controller) does it and I quite liked it. We can in the future use this to add options/refactor more to make it more clean.

Reply:
That's good enough for me.